### PR TITLE
Point groups

### PR DESCRIPTION
**Pull request by ewchow**
Hi @ankitrohatgi 👋 , hope you are well. Got a new feature here that will allow users to define related groups of points (such as error bar data). Let me know your thoughts, happy to make any changes.

Related to #71, #51, and #248.

### Summary
Point groups allow users to define groups of related data, such as standard deviation, standard error, or particular confidence intervals; and collect data points in tuples, in the order of the defined groups. This feature is available on XY plots, bar plots, polar diagrams, ternary diagrams, and images. Maps are not supported (I didn't think it would be useful). If no point groups are defined, the JSON export will not contain any additional key-value pairs. Unit tests have been added for the new infrastructure and a few pre-existing functions.

### How to use point groups
To use point groups, first create an axes of a supported axes type and an accompanying dataset.
#### Defining point groups
Under the `Dataset` menu on the left, there is a new `Edit Point Groups` button.

image

Clicking on this button will open a dialog box with 1 input field and several buttons. The input field (labelled `Group 0`) is the primary group. The primary group is the "root" of each tuple and can have a custom name.

image

To add new groups, click on the `Add New Group` button near the bottom. This will display a new input field labelled `Group 1`.

Click on the `Save` button when done.

Other than the primary group, each point group can be deleted from this dialog box. If a point group has associated data points, an additional dialog box will appear and the user will have the choice to delete the point group and all data points associated with the point group, or delete the point group without deleting the data points. If the latter is chosen, the data points will not be associated with any groups or tuples but remain in the dataset.

#### Collecting data points in tuples
Once point groups are defined, the `Add Point (A)` tool in manual extraction mode will display an additional UI on the right sidebar. This UI shows which point group and tuple the next data point will be associated with.

image

Data points will now be collected in the order of the defined point groups, in tuples.

For example, if we have defined the point groups "Mean", "Mean +1 SD", "Mean -1 SD"; the first data point added will be associated with "Mean", second with "Mean +1 SD", third with "Mean -1 SD", fourth with "Mean", etc.

Each time the last point group is filled on a tuple, the next data point will be associated with a new tuple as its primary point group. In the above example, the fourth data point will be associated with the "Mean" point group, but unlike the previous 3 data points, it will be associated with a new tuple.

The `Prev Group (,)` and `Next Group (.)` buttons allow users to skip back and forward, respectively, to tuples without all point groups filled. This occurs when a data point is deleted from a tuple or a new point group is defined after data points have been collected. Once a tuple is created, `Next Group (.)` also allows users to skip forward to skip point groups, or associate the next data point with a new tuple.

#### Deleting data points
When deleting a data point while point groups are in use, a dialog box will display. Users can either delete the selected data point, or delete all the data points in the tuple (this will delete the tuple as well).

#### Viewing data
Point group data will appear in the `View Data` dialog box. There are two new headings, Tuple and Group.

image

Tuple is a 0-indexed integer assigned to each tuple. All data points in the tuple will have the same tuple index.
Group is the point group name associated with the data point.

#### Bar plots
Bar plots have labels for each bar. When point groups are in use, only data points associated with the primary group will have a label displayed graphically.

image

When viewing the data through the `View Data` dialog box, all non-primary group data points will have the label of the primary group data point of its associated tuple in the first column. 

image

### CODE REVIEW

**Comment by billdenney** [comment]
I just did some testing of this, and it looks like it will fully address everything in #51, #71, and #248 for my use cases.

**Comment by ankitrohatgi** [comment]
Hi @ewchow, I finally decided to test this out a bit.

I see that datasets tied to `Map` axes are do not show the `Edit Point Groups` option. Is that intentional?

File: app/javascript/widgets/pointGroups.js [hunk 1]
```diff
@@ -0,0 +1,410 @@
+/*
```
Comment (ankitrohatgi) [inline app/javascript/widgets/pointGroups.js:1]: @ewchow - can you include this file in `templates/dev.html`?

**Comment by ankitrohatgi** [comment]
So, this works as described, but I found the interface a bit confusing unless I read the instructions here. Perhaps improving the text in the `Edit Point Groups` popup and/or adding some images might help? I'm concerned that a lot of people won't read the instructions very carefully and make a lot of errors or think this is broken.

**Comment by ewchow** [comment]
That's a good point. I'll update the text and include some images.

I could be wrong about this, but I didn't think there was a use case for Point Groups within `Map` axes, since the xy-coordinates are locations.

**Comment by ankitrohatgi** [comment]
Let's merge this in as-is. Improvements can come later.

**Pull request merged**